sql/plan/join.go

```diff
@@ -41,709 +34,171 @@ var useInMemoryJoins = shouldUseMemoryJoinsByEnv()
-		attribute.String("left", leftName),
-		attribute.String("right", rightName),
-	))
-
-	var inMemorySession bool
-	val, err := ctx.GetSessionVariable(ctx, inMemoryJoinSessionVar)
-	if err == nil && val == int8(1) {
-		inMemorySession = true
-	}
-
-	if mode == unknownMode {
-		if useInMemoryJoins || inMemorySession {
-			mode = memoryMode
-		}
-	}
-
-	cache, dispose := ctx.Memory.NewRowsCache()
-	if typ == JoinTypeRight {
-		r, err := right.RowIter(ctx, row)
-		if err != nil {
-			span.End()
-			return nil, err
-		}
-		return sql.NewSpanIter(span, &joinIter{
-			typ:               typ,
-			primary:           r,
-			secondaryProvider: left,
-			cond:              cond,
-			mode:              mode,
-			secondaryRows:     cache,
-			rowSize:           len(row) + len(left.Schema()) + len(right.Schema()),
-			dispose:           dispose,
-			originalRow:       row,
-			scopeLen:          scopeLen,
-		}), nil
-	}
-
-	l, err := left.RowIter(ctx, row)
-	if err != nil {
-		span.End()
-		return nil, err
-	}
-
-	return sql.NewSpanIter(span, &joinIter{
-		typ:               typ,
-		primary:           l,
-		secondaryProvider: right,
-		cond:              cond,
-		mode:              mode,
-		secondaryRows:     cache,
-		rowSize:           len(row) + len(left.Schema()) + len(right.Schema()),
-		dispose:           dispose,
-		originalRow:       row,
-		scopeLen:          scopeLen,
-	}), nil
 }
 
-// joinMode defines the mode in which a join will be performed.
-type joinMode byte
-
-const (
-	// unknownMode is the default mode. It will start iterating without really
-	// knowing in which mode it will end up computing the join. If it
-	// iterates the right side fully one time and so far it fits in memory,
-	// then it will switch to memory mode. Otherwise, if at some point during
-	// this first iteration it finds that it does not fit in memory, will
-	// switch to multipass mode.
-	unknownMode joinMode = iota
-	// memoryMode computes all the join directly in memory iterating each
-	// side of the join exactly once.
-	memoryMode
-	// multipassMode computes the join by iterating the left side once,
-	// and the right side one time for each row in the left side.
-	multipassMode
-)
-
-// joinIter is a generic iterator for all join types.
-type joinIter struct {
-	typ               JoinType
-	primary           sql.RowIter
-	secondaryProvider rowIterProvider
-	secondary         sql.RowIter
-	cond              sql.Expression
-
-	primaryRow sql.Row
-	foundMatch bool
-	rowSize    int
-
-	// scope variables from outer scope
-	originalRow sql.Row
-	scopeLen    int
-
-	// used to compute in-memory
-	mode          joinMode
-	secondaryRows sql.RowsCache
-	pos           int
-	dispose       sql.DisposeFunc
-}
-
-func (i *joinIter) Dispose() {
-	if i.dispose != nil {
-		i.dispose()
-		i.dispose = nil
+// Expressions implements sql.Expression
+func (j *JoinNode) Expressions() []sql.Expression {
+	if j.Op.IsLookup() || j.Op.IsDegenerate() {
```

Comment:
Comment on why `IsLookup()` isn't Expressioner in particular? I would have thought it would still evaluate the index lookup condition? (At least in some cases; how liberal can indexes be in their return values?)